prosessering-core/src/main/kotlin/no/nav/familie/prosessering/internal/TaskService.kt

```diff
@@ -173,6 +173,14 @@ class TaskService internal constructor(
         return taskLoggRepository.countByTaskIdAndType(taskId, Loggtype.PLUKKET)
     }
 
+    @Transactional
+    fun settTilManuellOppfølging(task: Task, årsak: String? = null): Task {
```

Comment:
Hvordan er det tenkt at dette skal virke? 

Hvis det er ønskelig å gjøre det i prosessering av task, så tror jeg dette vil feile. Man hadde kunnet lagt til en exception, eks `SettTilManuellOppfølgningException` med årsak som fanges opp og håndteres i `doFeilhåndtering`

I frontend kan man sette en task til manuell oppfølgning

Reply:
Ja, tanken er å gjøre det når vi prosesserer en task. Vi ønsker gjøre noe à la 
```
val status = oppdragClient.hentStatus(oppdragId)
if (status == KVITTERT_FUNKSJONELL_FEIL) {
    oppdaterIverksettResultat()
    settTaskTilManuellOppfølging()
} 
```
Snakket litt med Charlie om dette, men mulig det ikke vil fungere å oppdatere tasken mens den prosesseres, Stephen og jeg kjenner ikke dette biblioteket spesielt godt. Vi ønsker ikke å bruke frontend til dette, vil at det skal skje automatisk i koden. Mulig det er bedre å lage en custom exception og fange det opp i taskworkeren som du sier.

Reply:
Det du prøver å si er at så lenge vi ikke kaster en exception, vil `TaskWorker`en sette tasken til ferdig i post-prosesseringen, og `settTilManuellOppfølging` har ingen effekt?

Reply:
Jeg vil egentlig unngå å kaste en exception i det hele tatt. Jeg tror en `SettTilManuellOppfølgingException` blir en suboptimal løsning for appene som bruker task-rammeverket fordi man ikke vil at den skal oppføre seg som en exception, man vil bare sette tasken til manuell oppfølging. Så hvis man kaster `SettManuellOppfølgingException` i klientbiblioteket, må man passe på å unnta den fra evt. transaksjonshåndtering (som du også er inne på), fordi man egentlig bare vil trigge en custom kodeflyt i et eksternt bibliotek (i `doFeilhåndtering`). Det tror jeg vi fort mister kontroll på om x antall måneder, og jeg tror det blir vanskelig å forstå for nye utviklere.  

Metoden som @stephenramthun og jeg lagde i `TaskService` kaster jo ingen exception, men jeg tror du har helt rett i at det ikke vil fungere å kalle den direkte i vår app mens `doTask` kjører, fordi `TaskWorker` uansett vil kjøre `ferdigstill(task)` i etterkant som overskriver statusen til Ferdig. Ideelt sett hadde task-rammeverket hatt støtte for å la klienter skille på end-states som er OK og end-states som trenger oppfølging/trenger å fanges opp (derav ønsket om å sette tasken til manuell oppfølging). Da kunne `TaskWorker` hatt mer eksplisitt flyt à la 

```
val trengerVidereOppfølging = doTask(task)
...
if (trengerVidereOppfølging) {
 taskService.settTilManuellOppfølging(task)
} else {
   taskService.ferdigstill(task)
}
```
I stedet for skjult logikk i `doFeilhåndtering`. Men flyten jeg skisserer kommer til å treffe familie-appene ganske hardt, så tipper det ikke er ønskelig heller...

Reply:
Ja, jeg vil heller ikke ødelegge noe for andre brukere... Kan funke fint det du skisserer der, så kan vi override default implementasjon av `doTaskMedResultat` i dp-iverksett :)

Reply:
Tror også en `doTaskMedResultat` blir mest hensiktsmessig her - hvis vi forventer at `doTask` skal returnere noe fornuftig så kommer det til å bli mye feil rundt omkring :)

Reply:
Man har:
settTilManuellOppfølgning = true,
i annoteringen.

Da settes tasken til manuell oppfølging i stedet for feil.

Reply:
@stigebil ulempen med det er at man må kaste exception X ganger. Og transaksjonen blir rullet tilbake.
Det Martine ønsker er en mer kontrollert sett til manuell behandling

Reply:
Slik jeg tolket det på Martine var det ikke ønskelig med å rulle tilbake transaksjon i dette tilfelle.
Mitt første forslag var å bruke exception for dette, som jeg tenker burde dekke de fleste tilfeller :)

Det føles også litt rart at man skal sette en task til manuell oppfølgning uten at det tasken utfør ikke rulles tilbake. Det er kanskje det du tenker på? At man får en del kompleksitet ved rekjøring av tasken når den kanskje allerede har lagret ned noe..

Reply:
Det jeg tenker på er at når man har en Transactional, så kan transaksjon markeres som rollback only ved en exception hvis man ikke eksplisitt håndterer det. Dette gjelder også om man catcher exception og går videre. Man vil få Transaction marked as rollbackonly når tasken commites. Selvfølgelig kun et problem hvis det kastes en exception i tasken.

Men hvis man ønsker å commit tasken og sette til manuell oppfølging, så mener jeg at man fortsatt kan bruke settTilManuellOppfølgning = true. Tasken kaller en funksjon i en service med
`@Transactional(propagation = Propagation.REQUIRES_NEW)
`
Det som skjer i funksjonen commites i en egen transaksjon(Anbefaler at funksjonen er idempotent). Så kan resultatet av denne kjøringen avgjøres om tasken gikk bra. Eller om den gikk dårlig så kaster man en exception og den settes til manuell. Hvis man ikke ønsker flere rekjøringer så setter man maxAntallFeil = 1

Reply:
@stigebil Litt seint på ballen her, beklager det. Kan prøve å forklare hva vi vil og ikke vil gjøre her... Tasken det er snakk om er en `VentePåStatusFraØkonomiTask` i dp-iverksett som er en fork av familie-ef-iverksett. I en del tilfeller tar det lang tid å få kvittering fra OS, og hvis tasken ikke har fått OK-kvittering, kastes det en `TaskExceptionUtenStackTrace` og tasken rekjører. Her har vi satt `maksAntallFeil = 50` og `settTilManuellOppfølging = true` fordi det ofte tar litt tid før vi får kvittering fra OS og dermed må kjøre tasken en del ganger, og vi klarer å fange opp iverksettinger som ikke har fått kvittering innen rimelig tid. Dette fungerer stort sett veldig bra slik flyten og oppsettet er i dag. 

Det som derimot ikke fungerer så bra er de tilfellene der vi får en feilkvittering tilbake fra OS. I disse tilfellene hjelper det ikke å rekjøre tasken, og vi ønsker å oppdatere tilstanden i dp-iverksett med feil-statusen vi har fått fra OS og fange opp at vi må følge opp denne iverksettingen. Derav ønsket om å kunne sette tasken direkte til manuell oppfølging uten at den rekjøres og uten at transaksjonen rulles tilbake. Per i dag havner iverksettingene med feilkvittering inn i den samme flyten som iverksettingene som venter på kvittering, og det er uheldig blant annet fordi vi aldri får oppdatert tilstanden på iverksettingene med feilkvittering. Vi har et status-endepunkt som følgelig gir feil status for disse iverksettingene. 

Jeg synes det er en dårlig løsning å kaste exception i de tilfellene vi får feilkvittering fordi vi vil ikke rekjøre tasken og vi vil ikke rulle tilbake transaksjonen. Samtidig ønsker vi å beholde flyten slik den er i dag for de iverksettingene som venter på kvittering. Alternativet for vår del dersom vi ikke får endret task-rammeverket, er å sette `VentePåStatusFraØkonomi`-tasken til `Ferdig` når vi får en feilkvittering og ha en ny tasktype som kjører med jevne mellomrom og sjekker etter feilkvitteringer i databasen til dp-iverksett. Den nye tasktypen kunne nok kastet en exception hvis den finner feilkvitteringer, og vi kunne satt `settTilManuellOppfølging = true` og `maxAntallFeil = 1`. Da ville vi oppnådd nesten det samme, bortsett fra at iverksettinger potensielt kan ligge en stund uten at vi oppdager feilkvitteringen, avhengig av hvor ofte den kjøres.

Reply:
Jeg har ikke den totale oversikten over behovene deres, @marenger, men kunne man løst det på en annen måte? 

Å lage en egen "OppfølgingAvFeiledeØkonomiTask" dersom `VentePåStatusFraØkonomiTask` returnerer en tilstand som må håndteres på en annen måte?

Isåfall kan du i `onCompletion` sjekke på staten du har persistert og opprette en egen task-type for å håndtere de som er i en pussig tilstand? 

Da kan du vel sette den til `MANUELL_OPPFØLGING` direkte? 

Denne tasken blir jo trolig litt lik i innhold som `VentePåStatusFraØkonomiTask`, men det er iallefall mye lettere å "skille de ut" dersom man har en vaktordning og skal sjekke på hva er kritisk og hva er nede pga økonomi har "stengt"?

Reply:
@marenger Mitt forslag ruller ikke tilbake transaksjonen
```
@Service
@TaskStepBeskrivelse(
    taskStepType = "foo",
    beskrivelse = "Foo",
    maxAntallFeil = 3,
    settTilManuellOppfølgning = true,
)
class FooTask(
    private val fooService: FooService,
) : AsyncTaskStep {
    override fun doTask(task: Task) {
        val resultat = fooService.kjørOgKomitt()
        if (resultat == "TIL_MANUELL") error("Task settes til manuell")
    }
}

@Service
class FooService() {
    @Transactional(propagation = Propagation.REQUIRES_NEW)
    fun kjørOgKomitt(): String {
        return "TIL_MANUELL"
    }
}

```

Reply:
@stigebil Ulempen med den løsningen er at det finnes tilfeller der det kan være fint å sette tasken til manuellOppfølgning direkte, uten at den som i det der tilfellet prøver på nytt flere ganger først. 

At saker skjer i en egen transaksjon også bør gjøres veldig kontrollert då det kan ha **uønskede konsekvenser**. 

Forslagen tenker jeg er gode:
* Ny metode i interface (Johan)
* Kast exception og kjør oppdatering av resultat i egen transaction (Johan/Stig)
* Egen task med status MANUELL_OPPFØLGING (Charlie)

Heller nesten på den siste, tenker det er en grei løsning. Då får man oppdatert resultat, og lagt inn en ManuellOppfølgningstask som kan avvikshåndteres uten noen direkte konsekvenser.

Reply:
En egen task for å håndtere feilkvitteringer som settes til manuell oppfølging var vel strengt tatt mitt forslag, men samme det 😅 Jeg tenker at det er en helt grei løsning for vår del hvis dere ikke vil endre på `AsyncTaskStep`-interfacet. Innmaten i den tasken blir sikkert seendes litt rar ut, men flyten er ganske eksplisitt og enkel å følge, og det er det viktigste. Ellers tror jeg ikke det har så mye å si for oss om tasken sjekker etter feilkvitteringer selv eller om vi oppretter selve tasken når vi vet at vi har fått feilkvittering.

Reply:
Jeg tolket det som 2 ulike forslag